ext/mbstring/mbstring.c

```diff
@@ -4379,12 +4379,28 @@ MBSTRING_API int php_mb_check_encoding(const char *input, size_t length, const m
 		ZEND_ASSERT(out_len <= 128);
 		for (int i = 0; i < out_len; i++) {
 			if (wchar_buf[i] == MBFL_BAD_INPUT) {
-				return 0;
+				return false;
 			}
 		}
 	}
 
-	return 1;
+	return true;
+}
+
+static bool mb_check_str_encoding(zend_string *str, const mbfl_encoding *encoding)
+{
+	if (encoding == &mbfl_encoding_utf8) {
+		if (GC_FLAGS(str) & IS_STR_VALID_UTF8) {
+			return true;
+		}
+		bool result = php_mb_check_encoding(ZSTR_VAL(str), ZSTR_LEN(str), encoding);
+		if (result && !ZSTR_IS_INTERNED(str)) {
```

Comment:
is `IS_STR_VALID_UTF8` flag always computed when string is interned?

if yes, `php_mb_check_encoding` is not needed to be evaluated (just check if interned and return the flag)

if not, is there anything against doing so?

Reply:
From what I can tell, at least persistent interned strings must not be changed during requests, since these are stored in a true C global (and as such writing is not thread-safe). If OPcache is enabled, that basically holds for all interned strings.

Reply:
@cmb69 Thanks for that enlightening information! :+1: